client/cmd/dexc-desktop/app_darwin.go

```diff
@@ -0,0 +1,578 @@
+		webView.LoadRequest(req)
+		webView.SetNavigationDelegate_(cocoa.DefaultDelegate)
+	}
+
+	// WebView will execute this method when the page has loaded. We can then
+	// create a new window to avoid a temporary blank window. See:
+	// https://developer.apple.com/documentation/webkit/wknavigationdelegate/1455629-webview?language=objc
+	// NOTE: This method actually receives three argument but the docs said to
+	// expect two (webView and navigation).
+	addMethodToDelegate("webView:didFinishNavigation:", func(_ objc.Object /* delegate */, webView objc.Object, _ objc.Object /* navigation */) {
+		// Return early if we already created a window for this webview.
+		if !core.True.Equals(webView.Send(selIsNewWebview)) {
+			return // Nothing to do. This is just a normal window refresh.
+		}
+
+		// Overwrite the webview "selIsNewWebview" method.
+		webView.Class().AddMethod(selIsNewWebview, selFalse)
+
+		nOpenWindows.Add(1) // increment the number of open windows
+
+		// Create a new window and set the webview as its content view.
+		win := cocoa.NSWindow_Init(core.NSMakeRect(0, 0, float64(width), float64(height)), cocoa.NSClosableWindowMask|cocoa.NSTitledWindowMask|cocoa.NSResizableWindowMask|cocoa.NSFullSizeContentViewWindowMask|cocoa.NSMiniaturizableWindowMask, cocoa.NSBackingStoreBuffered, false)
+		win.SetTitle(appTitle)
+		win.Center()
+		win.SetMovable_(true)
+		win.SetContentView(webkit.WKWebView_fromRef(webView))
+		win.SetMinSize_(core.NSSize{Width: 600, Height: 600})
+		win.MakeKeyAndOrderFront(nil)
+		win.SetDelegate_(cocoa.DefaultDelegate)
+	})
+
+	// Add custom selectors to the app delegate since there are reused in
+	// different menus. App delegates methods should be added before NSApp is
+	// initialized.
+	addMethodToDelegate(selOpenLogs, func(_ objc.Object) {
+		logDirURL, err := app.FilePathToURL(logDir)
+		if err != nil {
+			log.Errorf("error constructing log directory URL: %v", err)
+		} else {
+			openURL(logDirURL)
+		}
+	})
+
+	addMethodToDelegate(selNewWindow, func(_ objc.Object) {
+		windows := cocoa.NSApp().OrderedWindows()
+		len := windows.Count()
+		if len < uint64(maxOpenWindows) {
+			createNewWebView()
+		} else {
+			// Show the last window if maxOpenWindows has been exceeded.
+			winObj := windows.ObjectAtIndex(len - 1)
+			win := cocoa.NSWindow_fromRef(winObj)
+			win.MakeMainWindow()
+		}
+	})
+
+	// MacOS will always send the "applicationShouldTerminate" event when an
+	// application is about to exit so we should use this opportunity to
+	// cleanup. See:
+	// https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428642-applicationshouldterminate?language=objc
+	addMethodToDelegate("applicationShouldTerminate:", func(s objc.Object) core.NSUInteger {
+		cancel()                 // no-op with clean rpc/web server setup
+		wg.Wait()                // no-op with clean setup and shutdown
+		shutdownCloser.Done(log) // execute defer statements
+		shutdownCloser.Success()
```

Comment:
> This just isn't how ErrorCloser is intended to be used. 

Yh. But it fits the use case and I didn't see the need to create a new type for this.

> Besides the fact that you'd be calling Done twice in this case (defer above)

The defer functions will `never execute`. The app can be terminated from different places (main menu, systray, dock or via Activity Monitor) after which macOS will execute the app delegate's [applicationShouldTerminate:](https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428642-applicationshouldterminate?language=objc)` method. This gives us the opportunity to clean up (i.e. to execute functions we could have deferred).

[From the docs](https://developer.apple.com/documentation/appkit/nsapplication/1428417-terminate?language=objc): 
> Don’t bother to put final cleanup code in your app’s main() function—it will never be executed. If cleanup is necessary, perform that cleanup in the delegate’s [applicationWillTerminate:](https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428522-applicationwillterminate?language=objc) method.